File: vercors/src/main/java/vct/col/ast/IfStatement.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import hre.ast.MessageOrigin;

import java.util.*;

public class IfStatement extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  public IfStatement(){
  }
  
  public IfStatement(ASTNode cond,ASTNode true_branch,ASTNode false_branch){
    addClause(cond,true_branch);
    if(false_branch!=null){
      addClause(else_guard,false_branch);
    }
  }
  
  

  public static final ASTNode else_guard=new ConstantExpression(true,new MessageOrigin("else guard"));

  private static class Case {
    public ASTNode guard;
    public ASTNode  statement;
  }
  
  private ArrayList<Case> cases=new ArrayList<Case>();
  
  public void addClause(ASTNode guard,ASTNode s){
    Case clause=new Case();
    clause.guard=guard;
    if (guard != else_guard) {
      guard.setParent(this);
    }
    clause.statement=s;
    s.setParent(this);
    cases.add(clause);
  }

  public int getCount(){ return cases.size(); }

  public ASTNode getGuard(int i){ return cases.get(i).guard; }
  
  public ASTNode getStatement(int i){ return cases.get(i).statement; }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
 

}

File: vercors/src/main/java/vct/col/ast/IfStatement.scala
package vct.col.ast

import hre.ast.MessageOrigin;
import vct.col.util.VisitorHelper

object IfStatement {
  def elseGuard = new ConstantExpression(true, new MessageOrigin("else guard"))
}

class IfStatement extends ASTNode with VisitorHelper  {
  private[this] val cases = new java.util.ArrayList[(ASTNode,ASTNode)]()
  
  def getCount = cases.size
  def getGuard(i:Int) = cases.get(i)._1
  def getStatement(i:Int) = cases.get(i)._2
  
  def addClause(guard:ASTNode, stmt:ASTNode) : Unit = {
    stmt.setParent(this)
    if (guard != IfStatement.elseGuard) {
      guard.setParent(this)
    }
    cases.add(new Tuple2(guard, stmt))
  }
  
  def this(cond:ASTNode, truebranch:ASTNode, falsebranch:ASTNode) = {
    this()
    addClause(cond, truebranch)
    if (falsebranch != null) {
      addClause(IfStatement.elseGuard, falsebranch)
    }
  }
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}
